core/src/types/storage.rs

```diff
@@ -1151,6 +1154,147 @@ pub struct PrefixValue {
+        }
+    }
+}
+
+/// Draining iterator over a queue of Ethereum events.
+pub struct EthEventsQueueIter<'queue, E> {
+    current_nonce: Uint,
+    queue: &'queue mut InnerEthEventsQueue<E>,
+}
+
+impl<E> Drop for EthEventsQueueIter<'_, E> {
+    fn drop(&mut self) {
+        // on drop, we commit the nonce of the next event to process
+        if self.queue.next_nonce_to_process < self.current_nonce {
+            self.queue.next_nonce_to_process = self.current_nonce;
+        }
+    }
+}
+
+impl<E: GetEventNonce> Iterator for EthEventsQueueIter<'_, E> {
+    type Item = E;
+
+    fn next(&mut self) -> Option<E> {
+        let nonce_in_queue = self.queue.peek_event_nonce()?;
+        if nonce_in_queue == self.current_nonce {
+            self.current_nonce = self
+                .current_nonce
+                .checked_increment()
+                .expect("Nonce overflow");
+            self.queue.pop_event()
+        } else {
+            None
+        }
+    }
+}
+
+impl<E: GetEventNonce> InnerEthEventsQueue<E> {
+    /// Retrieve the next events to be processed, if any.
+    ///
+    /// This decision is based on the nonce of the next event to be
+    /// processed by the ledger, and the nonce of the event that was just
+    /// confirmed (i.e. achieved a quorum of votes behind it).
+    pub fn get_next_events(
+        &mut self,
+        latest_event: E,
+    ) -> EthEventsQueueIter<'_, E>
+    where
+        E: std::fmt::Debug,
+    {
+        let event_nonce = latest_event.get_event_nonce();
+        if hints::unlikely(self.next_nonce_to_process > event_nonce) {
+            unreachable!(
+                "Attempted to replay an Ethereum event: {latest_event:#?}"
+            );
+        }
+
+        self.push_event(latest_event);
+
+        EthEventsQueueIter {
+            current_nonce: self.next_nonce_to_process,
+            queue: self,
+        }
+    }
+
+    /// Provide a reference to the earliest event stored in the queue.
+    #[inline]
+    fn peek_event_nonce(&self) -> Option<Uint> {
+        self.inner.front().map(GetEventNonce::get_event_nonce)
+    }
+
+    /// Push a new transfer to Namada event to the queue.
+    #[inline]
+    fn push_event(&mut self, new_event: E)
+    where
+        E: std::fmt::Debug,
+    {
+        self.inner
+            .binary_search_by_key(
+                &new_event.get_event_nonce(),
+                |event_in_queue| event_in_queue.get_event_nonce(),
+            )
+            .map_or_else(
+                |insert_at| {
+                    tracing::debug!(?new_event, "Queueing Ethereum event");
+                    self.inner.insert(insert_at, new_event)
+                },
+                // the event is already present in the queue... this is
+                // certainly a protocol error
+                |_| {
+                    hints::cold();
+                    unreachable!(
+                        "An event with an identical nonce was already present \
```

Comment:
The doc string and the unreachable message imply different things. I think the unreachable message is the correct one.